### ISSUE TITLE
[Feature] Language switcher

### ISSUE THREAD

**Issue by czukowski** [Ryochan7/DS4Windows#169]
I know this is not a very important feature and in fact, probably very few people would use it, but I decided to implement it because I like to be able to change the language in the applications I use and I noticed some issues in layouts of certain localized forms. Last but not least, I also wanted to refresh my memory of .net development and this seemed as a good way for it :)

https://github.com/czukowski/DS4Windows/commits/language-switch

I created a new custom control that combines a label and a combo box and fills it with the available languages. I also added some code to change application settings with selected values from the combo box. I haven't added the control to `DS4Form` though, because when I tried, it has messed up `DS4Form.resx` file beyond any hope of successful merging as the elements seemed to have been re-ordered randomly. So I'm not sure how to proceed there.

In order to add the new control to the Settings tab:

1. Drag LanguagePackComboBox from Toolbox to an appropriate place in `DS4Form` designer
2. Add event handler for `SelectedValueChanged` to `languagePackComboBox1_SelectedValueChanged` method
3. Optionally, set the control's `InvariantCultureText` property to something like "None (English)" in resource editor, otherwise the 1st combo box element will be empty

The application must be restarted before the language change takes effect. I've seen examples how to update form controls with another culture somewhere, but that seemed to require larger changes, so I decided to maybe revisit it later.

One more issue I've encountered is that logged messages appear localized with the system culture no matter what. Both `Thread.CurrentThread` and `Program.controlThread` are changing their `CurrentUICulture` property on `DS4Form` load, and I've also tried to change `CurrentCulture` as well, and also hardcoding the values at the 1st line of `Program.Main` method (ie before the application creates new threads), still to no avail. I can see that `ControlService` uses static calls to resources (`Properties.Resources.*`) and not the resource manager, but so are some message boxes in `DS4Form` (for example the one saying DS4Windows is up to date) and they show the messages in correct language. I'm clearly missing something, but what?

**Comment by Ryochan7** [comment]
A language switcher feature might be useful. My foreign language knowledge is limited so I would probably not use it too much; I know some elementary Japanese and the little bit of Spanish that I remember learning in high school. I will have to take a look at your fork soon and see how well it performs.

Regarding the way Resources are handled, that is the method that Visual Studio gives a project for free. Behind the scenes, the Resources class creates a static instance of the ResourceManager and uses that to obtain any bundled resources. The only other C# project that I hack around with on occasion is Touchmote and it uses the same approach. The message box that you referred to uses the resource manager to obtain the necessary string to display.

Changing the language without restarting the application would probably require grabbing an event of some kind and refreshing all the controls for open forms.

**Comment by czukowski** [comment]
Found an answer at StackOverflow about culture in child processes and it seemed to work, log messages are now localized.

https://stackoverflow.com/a/17701938/1570537
czukowski/DS4Windows@60916e10800b9b1cc97b2e0dd5d73c0ba4a63784

The changes should have negligible impact on speed, essentially it's just a culture is being set during startup as seen in the commit above, the only thing is that it originates from DS4Form since it's the one that loads the settings. Now that it's not necessary to modify the control thread anymore, it may be a better idea to move this code back to DS4Form class. In addition to that I'll need to make the new dialog box a bit prettier and its text localizable, then I can open a pull request if you're open to this kind of features.

A problem still remains with merging of resx files, I have merged some of your new upstream commits while I was playing with the language picker and the resx files were drastically different even though I've only added a control to the form. I've eventually given up and opted to use the upstream file and re-add the control. There must be a way people do it right, it's not possible that only one developer gets to edit forms at a time :)

**Comment by Ryochan7** [comment]
The changes are looking good so far. There doesn't seem to be a performance difference so that is a relief; editing almost anything in the project can be unnerving with some minor code changes drastically changing performance. One problem that I do see currently is that the approach used to find localization assemblies will not work for public Release builds.

The current approach uses the standard .NET practice of looking for a language sub-folder in the main executable directory. The App.config file of DS4Windows has been updated so .NET will use a Lang sub-folder as a secondary search path for localization folders. Localization files in recent builds of DS4Windows are included in the Lang folder. Hopefully there is a different way to find the relevant language assemblies especially if it is one that can rely on .NET to handle the dirty work of finding any included localization assemblies.

As for the resx files, Visual Studio heavily changes the structure of those files any time that a form is edited. As long as no real conflict occurs then I would not worry about it. The resulting diff of including the LanguagePackComboBox in DS4Form looks fine on my end.

**Comment by czukowski** [comment]
You're right about the language assemblies, I figured .NET uses some kind of magic to find those that I did not understand, my implementation was based on one of the examples I've found and it sure could be improved, I'll try to dig some more. I also don't like the fact that it's listed some languages like for example `de` for having a seemingly unrelated DLL inside (while DS4Windows.resources.dll is located in `de-DE` subfolder).

As for the resx files, I've found a reference to ResX Manager extension for Visual Studio that is supposedly capable of reordering XML elements in the .resx files on save in a pre-defined order, which should help a big deal with the diffs. I've checked it out just briefly, but perhaps I haven't done something right (just tried edit some strings back and forth in order to trigger save and re-order, without actually adding any new strings or translations), and couldn't reproduce the advertised effect, at least not in full. I'll have another go at this when I add some localization for the new control and dialog box.
